packages/repository/test/unit/repositories/constraint-utils.unit.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {
+  FilterBuilder,
+  Filter,
+  Fields,
+  Where,
+  WhereBuilder,
+} from '../../../src/query';
+import {AnyObject} from 'loopback-datasource-juggler';
+import {
+  constrainFilter,
+  constrainWhere,
+} from '../../../src/repositories/constraint-utils';
+
+describe('constraint utility functions', () => {
+  let originalFilter: Filter = {};
+  let originalWhere: Where = {};
+  const constraint: AnyObject = givenAWhereConstraint();
+
+  before(() => {
+    originalFilter = givenAFilter({a: true}, {x: 'x'}, 5);
+    originalWhere = givenAWhere();
+  });
+  context('constrainFilter', () => {
+    it('applies a where constraint', () => {
+      const result = constrainFilter(originalFilter, constraint);
+      expect(result).to.deepEqual({
+        fields: {a: true},
+        limit: 5,
+        where: {x: 'x', id: '5'},
+      });
+    });
+    it('applies a filter constraint with where object', () => {
+      const validFilterConstraint = givenAFilter(
+        undefined,
+        {id: '10'},
+        undefined,
+      );
+      const result = constrainFilter(originalFilter, validFilterConstraint);
+      expect(result).to.deepEqual({
+        fields: {a: true},
+        limit: 5,
+        where: {x: 'x', id: '10'},
+      });
```

Comment:
We should have only one test that's verifying whether `constraintFilter` changes or preserves filter properties we are not changing in the test (`fields`, `limit`, etc.).

In your current test design, if we add another property to `originalFilter`, let's say `include`, we would have to update all tests to check `include` too.

What to do instead: use one of the `contain*` matchers to verify only the property this test is interested in.

```ts
expect(result).to.containEql({
  where: /* expected where */
});
```

A full test case as I would write it, taking into my consideration the previous comment too.

```ts
const input = givenAFilter({where: {x: 'x'}});
const constraint = givenAWhereConstraint({id: '10'});
const result = constrainFilter(input, constraint);
expect(result).to.containEql({
  where: {x: 'x', id: '10'}
});
```

Reply:
Done :-).